### ISSUE TITLE
FeatureFinderMetabo uses start of feature as centroid RT

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#2777]
FeatureFinderMetabo (FFM) doesn't adhere to the convention for feature (centroid) positions in OpenMS. From the documentation of the "Feature" class (http://ftp.mi.fu-berlin.de/pub/OpenMS/develop-documentation/html/classOpenMS_1_1Feature.html):
> By convention in OpenMS, the position of a feature is defined as maximum position of the model for the retention time dimension and the mass of the monoisotopic peak for the m/z dimension.

FFM assigns the start of the RT range as the RT position of the feature, not the RT value of the elution peak apex. This means when a feature is represented only by its centroid (e.g. in a consensus map), its position is shifted away from where it would be expected based on the LC-MS signal.

**Comment by timosachsenberg** [comment]
Sounds like a nasty bug. Can you trace it down to a line?

**Comment by hroest** [comment]
definitely does not look like the intention: https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp#L894

```
      Feature f;
      f.setRT(feat_hypos[hypo_idx].getCentroidRT());
      f.setMZ(feat_hypos[hypo_idx].getCentroidMZ());
```

which calls `MassTrace::getCentroidMZ` which is a cached variable that gets updated if you call the right functions on `MassTrace`, e.g. `MassTrace::updateWeightedMeanMZ` among others. But it looks like all mass traces are update before they are pushed into the output vector: https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FILTERING/DATAREDUCTION/MassTraceDetection.cpp#L537

**Comment by hendrikweisser** [comment]
Apologies, my bug report was definitely too general - I've now also seen features where everything looks okay. I'll check what went wrong in the problematic examples.

**Comment by timosachsenberg** [comment]
@hendrikweisser I think I recall that FFM starts at the apex and then tries to extend the mass trace on both sides. Sometimes it fails in one direction which would result in the apex beeing placed on the lowest/highest RT. Could this be the case?

**Comment by hendrikweisser** [comment]
@timosachsenberg: Yes, that could be the case. However, I definitely see examples where the apex is close to the beginning (in RT) of the detected feature, but not *at* the beginning, and the centroid RT is assigned incorrectly. In other cases, it looks like the elution profile detection, which splits some mass traces into multiple features, might play a role.

**Comment by hendrikweisser** [comment]
I'm looking at an example now where an elution peak from a 13C mass trace was cut into three parts. The middle bit (containing the apex) was joined together with the 12C trace and the 2*13C trace into a feature (good!). The parts on the left and right were turned into separate features, with centroids closest to the middle - which is where the max. intensity is, but still not ideal.

**Comment by hroest** [comment]
one question is whether the elution profile detection recomputes centroid m/z and RT values or not? what happens when you run it without the elution profile detection?

**Comment by hendrikweisser** [comment]
I'm still testing different parameter combinations, trying to find one that works well for my data. I've identified two (related) problems with the elution profile detection:
1. There's no single value of the "chrom_fwhm" (expected chromatographic peak width) parameter that works well for  high *and* low intensities. It either leads to splitting of broad, high-intensity peaks (see 2.) or fails to detect narrow, low intensity peaks. (Compare the "matched filter" approach implemented e.g. in XCMS, which uses a set of different peak widths.)
2. There's an unfortunate tendency to split high-intensity peaks into two halves, especially (but not exclusively) if they flatten out on top.

**Comment by jpfeuffer** [comment]
@hendrikweisser Do you remember if you recently still found weird behavior regarding this? Should we rename the issue maybe?

**Comment by timosachsenberg** [comment]
## Investigation

The report is real and still reproduces, but only under `epd:width_filtering = off` or `auto`. With the tool's built-in default (`fixed`) the reported RT is the elution apex, exactly, in every feature I checked. Three separate defects sit behind this thread.

### Where the RT comes from

`FeatureFindingMetabo.cpp:1036` → `iso_pattern_[0]->getCentroidRT()`, i.e. the cached `centroid_rt_` of the monoisotopic mass trace. Only two places write it:

* `ElutionPeakDetection.cpp:453,536` → `updateSmoothedMaxRT()` — RT of the max **smoothed** intensity (the apex, per the `Feature` convention)
* `MassTraceDetection.cpp:664` → `updateWeightedMeanRT()` — intensity-weighted **mean** RT, used only when `epd:enabled=false`

@hroest, to your 2017 question: EPD does recompute them. Split sub-traces get `updateSmoothedMaxRT` + `updateWeightedMeanMZ` + `updateWeightedMZsd`; unsplit traces get `updateSmoothedMaxRT`.

### Defect 1 — border apexes leak through when the width filter is off (the reported symptom)

`MassTrace::estimateFWHM()` returns **0** whenever the apex is the first or last point (`MassTrace.cpp:181`). Two things produce such traces: EPD splitting a peak at a minimum (@hendrikweisser's "left and right leftovers"), and MassTraceDetection's greedy seeding, where extension in one direction stops immediately because those peaks were already claimed by a more intense trace.

With `width_filtering=fixed` these are dropped (`FWHM 0 < min_fwhm`). With `off` or `auto` they become features whose RT is literally the first scan of their trace — and, because `computeFwhmArea()` short-circuits on `fwhm_start_idx_ == fwhm_end_idx_ == 0`, whose **intensity is 0**.

`src/tests/topp/FeatureFinderMetabo_6_input.mzML`, `noise_threshold_int=10`, `chrom_fwhm=5`:

| width_filtering | features | RT at first point | RT at last point | intensity == 0 |
|---|---|---|---|---|
| `fixed` (tool default) | 267 | 0 (0.0%) | 0 (0.0%) | 0 |
| `auto` | 534 | 200 (39.0%) | 69 (13.5%) | 269 |
| `off` | 589 | 223 (39.4%) | 76 (13.4%) | 299 |

Concrete case: trace `T14.1`, a 3-point fragment split off the leading edge of a 28-point trace spanning 2600.65–2654.22 s. Reported RT = **2600.65 s**, parent trace centroid = 2634.84 s → **34 s** off. Median shift over all 299 border-apex traces: 6.4 s, p90 20.4 s.

`auto` does not help, because `filterByPeakWidth` cuts at the 5% quantile of the FWHM distribution — and when more than 5% of traces have FWHM 0, that quantile *is* 0.

Note that `src/tests/topp/FeatureFinderMetabo.ini` itself ships `width_filtering=off`, so TOPP tests 1 and 5 run in the affected mode.

### Defect 2 — `updateWeightedMeanRT()` drops the first peak (`MassTrace.cpp:455`)

```cpp
double rt_before = trace_peaks_[0].getRT();
for (auto l_it = trace_peaks_.begin() + 1; ...)   // peak 0 never contributes
```

Right-Riemann weighting: peak 0 enters neither the numerator nor the denominator. For a **two-point trace the result is unconditionally the RT of the second point**, whatever the intensities — `[(100 s, I=1000), (105 s, I=10)]` gives a centroid of 105 s. On `FeatureFinderMetabo_1_input.mzML` that is 171 of 171 two-point traces; for traces with ≥3 points the median relative centroid position is 0.613 instead of ~0.5. This is the `epd:enabled=false` path (10.6% of features land on their last point on dataset 6). `MassTrace_test.cpp:764` hardcodes the biased value `155.214671250425`; mid-interval (trapezoidal) weights over all points give `155.209843462072`.

### Defect 3 — convention mismatch in the no-EPD path

`Feature` documents the RT position as "maximum position of the model". Without EPD the value is a weighted *mean*. Either that path needs a max-based update (there is no raw-max equivalent of `updateSmoothedMaxRT`), or the documentation should say so.

For @hendrikweisser's "apex close to the beginning but not at it" observation: with `fixed`, the feature RT equals the raw apex to the last digit in 190/190 features on dataset 1. Relative positions as low as 0.047 come from *lopsided trace boundaries* (MTD extended far in one direction), not from a wrong centroid.

### Provenance

Numbers were produced with the released pyOpenMS 3.5.0 wheel. Diffing `develop` against `release/3.5.0` shows `updateWeightedMeanRT` / `updateSmoothedMaxRT` / `estimateFWHM`'s border behaviour and the EPD/MTD algorithms unchanged apart from `String`→`std::string`, logging and an IM guard, so the mechanisms carry over; exact feature counts on `develop` will differ slightly since FFM gained `isotope_rt_overlap_reference` / `min_isotope_rt_overlap` from #4483.

### Suggested resolution

1. Reject traces whose apex sits on the first or last point in `ElutionPeakDetection`, regardless of `width_filtering` — this removes the reported symptom at its source and also stops the intensity-0 features. No effect on `width_filtering=fixed`.
2. Fix `updateWeightedMeanRT` to use mid-interval weights so peak 0 counts.

@jpfeuffer — re your 2022 question, renaming to something like *"FFM: features with border apexes get RT at trace edge and intensity 0 when `epd:width_filtering` is not `fixed`"* would match what actually happens. A PR for 1 and 2 follows.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by timosachsenberg** [comment]
Correction to my comment above, plus the PR: https://github.com/OpenMS/OpenMS/pull/10225

I had measured with the released pyOpenMS 3.5.0 wheel. Having since built `develop` and re-run the TOPP tests, one claim in that comment is wrong: I said FeatureFinderMetabo emits these as features with intensity 0 when `epd:width_filtering` is `off` or `auto`. It does not, because `src/topp/FeatureFinderMetabo.cpp:133` already erases zero-intensity features:

```cpp
// filter features with zero intensity (this can happen if the FWHM is zero (bc of overly skewed shape) and no peaks end up being summed up)
auto intensity_zero = [&](Feature& f) { return f.getIntensity() == 0; };
feat_map.erase(remove_if(feat_map.begin(),feat_map.end(),intensity_zero),feat_map.end());
```

That was added after 3.5.0. So on `develop`, with the default `mtd:quant_method=area`, `TOPP_FeatureFinderMetabo_1` writes 190 features whether or not the flank traces are filtered — the assembly still produces 447 (the "Output features: 447" log line is printed before the erase and overstates the count), and 257 of them are discarded on the way out.

What does survive on `develop`, measured with the built tool:

| configuration | features | RT at first point | RT at last point | intensity == 0 |
|---|---|---|---|---|
| FFM, `quant_method=area`, `width_filtering=off` | 190 | 0 | 0 | 0 |
| FFM, `quant_method=max_height`, `width_filtering=off` | 447 | 115 (25.7%) | 142 (31.8%) | 0 |
| MassTraceExtractor, `width_filtering=off` | 449 | 115 (25.6%) | 142 (31.6%) | 257 |

With `max_height` the flank's intensity is not 0, so the erase does not catch it and the feature is written with its RT on the edge of its own RT range. MassTraceExtractor has no such filter at all. Everything else in the investigation stands — the trace-level numbers, the mechanism, and the `updateWeightedMeanRT` bug, which is unaffected by any of this.

The PR removes the flanks in `ElutionPeakDetection` instead of relying on the zero-intensity erase, so the `max_height` and MassTraceExtractor cases are fixed too. The erase stays, since it still applies when `epd:enabled` is false.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by timosachsenberg** [comment]
A second correction, and the more important one. My "border apex means the trace is a flank" claim was wrong for the dominant case, and the first version of #10225 removed far too much as a result.

**What those traces actually are: peaks sampled too sparsely, not flanks.** `MassTraceExtractor_1_input.mzML` has a median scan spacing of 1.77 s against `chrom_fwhm=5 s` — about **2.8 scans per peak**. `MassTrace::estimateFWHM()` needs at least one point on each side of the apex to bracket the half maximum, so with ~3 points the maximum routinely lands on an end point and the function returns 0. That is a sampling property, not a defect in the trace.

Evidence that they are real signal:

* For **84 of the 85** multi-point cases, the scan just beyond the trace edge contains no peak at that m/z at all. They are not cut short by a neighbouring trace having claimed the peaks — only 1 was.
* **255 of 257** were never split by elution peak detection; they came out of `MassTraceDetection` that way. 57% of its output has the apex on an end point, before EPD runs at all.
* 53 of the 86 multi-point ones are not monotonic — they have peak shape.
* Median apex intensity 3208 against `noise_threshold_int=10`, and **15.9% of the total summed intensity**.
* Only 6 of 256 sit at the RT boundary of the run, so it is not an acquisition-edge effect.

So the filter I first proposed ("discard any trace whose apex is on its first or last point") deletes a large fraction of genuine, quantifiable peaks on data whose scan rate is close to the peak width. That was the wrong fix.

A narrowed version, rejecting only fragments whose apex sits on a boundary shared with another fragment of the same split trace, was also wrong and has since been removed from #10225 entirely — see the review thread there. Every cut is at a minimum and the split loop keeps that minimum as the last point of the left fragment, so the fragment after it starts one sample past the minimum and its first sample is normally a genuine maximum. Measured over the four test inputs, **107 of 107** fragments that rule rejected had an apex that is a local maximum of the parent trace.

**Correction to what follows (edited).** This comment originally said the over-splitting comes from `findLocalExtrema` clamping the split minimum to `>= 1.0`. That is wrong. `min_int = max(1.0, smoothed[min])` raises the *denominator* of the ratio test, making a split harder rather than easier, and an instrumented run over `FeatureFinderMetabo_6_input.mzML` — 1260 candidate minima, 250 accepted — shows the clamp was active in **0** of the accepted splits.

The two things this leaves open, both separate from #10225:

1. For an under-sampled peak the FWHM stays 0, so area-based quantification of that trace returns 0. `FeatureFinderMetabo` papers over it by erasing zero-intensity features (`src/topp/FeatureFinderMetabo.cpp:133`); `MassTraceExtractor` does not, and writes them with intensity 0. Estimating a one-sided FWHM in `estimateFWHM()` would fix both properly, but it changes quantification for every consumer (FFM, MassTraceExtractor, PeakPickerIM, FLASHDeconv) and deserves its own PR.
2. The over-splitting itself, which is what @hendrikweisser actually reported. The operative criterion is the hard-coded `left_max_int / min_int >= 2.0 && right_max_int / min_int >= 2.0`: a valley only has to fall below half of both neighbouring maxima for the trace to be cut. Over the 250 accepted splits above, the depth runs from exactly 2.00 upwards with a median of 2.71 — **at the median the valley sits at 37% of the smaller of the two peaks**. That matches "an unfortunate tendency to split high-intensity peaks into two halves, especially (but not exclusively) if they flatten out on top" from the thread above. Being looked at separately.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by timosachsenberg** [comment]
The RT-centroid half of this issue is fixed and merged (#10225, squashed as 158b899). This is the write-up of the over-splitting half, which is what @hendrikweisser actually reported and is still open.

## The suppression window collapses to ±1 scan

`ElutionPeakDetection::detectElutionPeaks_` derives everything from one quantity:

```cpp
double scan_time(mt.getAverageMS1CycleTime());
Size win_size = std::ceil(chrom_fwhm_ / scan_time);   // line 389
smoothData(mt, static_cast<Int>(win_size));           // frame_length = max(3, win_size), line 585
findLocalExtrema(mt, win_size / 2, maxes, mins);      // line 412 -- integer division
```

`findLocalExtrema` accepts a point as a maximum when nothing within `±num_neighboring_peaks` is higher. On `FeatureFinderMetabo_6_input.mzML` the median MS1 cycle time is **3.17 s**, so with the default `chrom_fwhm = 5 s`:

* `win_size = ceil(5 / 3.17) = 2`
* `num_neighboring_peaks = 2 / 2 = **1**`
* Savitzky-Golay `frame_length = max(3, 2) = **3**`, the smallest the filter accepts

So the trace is barely smoothed, and a point is declared an elution-peak maximum if it merely beats its immediate neighbour on each side. Any three-point wiggle is a peak. `MassTraceExtractor_1_input.mzML` (cycle 2.77 s) gives exactly the same 2 / 1 / 3.

Instrumenting `findLocalExtrema` to log the pairs of maxima it actually compares, on `FeatureFinderMetabo_6_input.mzML`: among the pairs it accepts as split points, the **median separation between the two maxima is 3 scans, and 48% are 1-2 scans apart** — two "peaks" declared within one expected peak width. Some are fully degenerate, e.g. a pair at indices 3 and 4 with the chosen minimum equal to the left maximum:

```
lm=3 min=3 rm=4  profile [13155, 21234]
```

Only after that do the split criteria run, and the `left_max_int / min_int >= 2.0 && right_max_int / min_int >= 2.0` test is easy to satisfy between two adjacent scans on a noisy edge.

## Consequence: `chrom_fwhm` is inert over most of its range

Because of the integer division, `chrom_fwhm` does nothing across the values anyone would set for LC peaks. On `FeatureFinderMetabo_6_input.mzML` (cycle 3.17 s, `epd:width_filtering=fixed`), varying only `algorithm:common:chrom_fwhm`:

| chrom_fwhm | win_size / half-window | features |
|---|---|---|
| 1 s | 1 / 0 | 239 |
| 2 s | 1 / 0 | 239 |
| 5 s | 2 / 1 | 239 |
| 10 s | 4 / 2 | 215 |
| 20 s | 7 / 3 | 181 |

1, 2 and 5 produce **identical output**. `MassTraceExtractor_1_input.mzML` behaves the same way: 1, 2, 3 and 5 all give 449 features, and only 10 moves it (447).

That is a direct match for the first complaint in this thread — *"There's no single value of the chrom_fwhm parameter that works well for high and low intensities"*. For a large part of its range it is not working at all, so no value of it could have helped.

At a half-window of 0 (whenever `scan_time >= chrom_fwhm`) the neighbour loop never executes:

```cpp
Size start_idx(0);
if (ref_idx > num_neighboring_peaks) { start_idx = ref_idx - num_neighboring_peaks; }
Size end_idx = ref_idx + num_neighboring_peaks;   // == ref_idx when the half-window is 0
...
for (Size j = start_idx; j < end_idx; ++j)        // start_idx == end_idx -> body never runs
```

`real_max` stays `true`, so every unused point above zero is accepted as a maximum. On the files above the downstream `min_dist` and valley tests limit the damage, so I have not observed a blow-up — but the guard is not doing anything in that regime.

## Effect of widening it

Prototype on branch `claude/brave-babbage-n9wo35-split` (commit `a271917`), which makes the half-window configurable as a multiple of `win_size` with a floor, both defaulting to current behaviour. The `x0.5 floor 1` row reproduces every reference file exactly, which is the check that the reimplementation is faithful — note `win_size / 2` floors, so `ceil` here silently shifts the baseline:

| half-window | FFM_1 | FFM_4 | FFM_6 | MTE_1 | MTE_4 |
|---|---|---|---|---|---|
| x0.5 floor 1 (current) | 190 | 4 | 239 | 449 | 261 |
| x0.5 floor 2 | 190 | 4 | 235 | 449 | 256 |
| x1.0 floor 1 | 189 | 4 | 233 | 448 | 254 |
| x1.0 floor 3 | 189 | 4 | 227 | 448 | 248 |

Widening to the full FWHM in scans removes roughly 2.5% of features. Fewer features means more traces assembled per feature, which the FeatureFinderMetabo documentation calls the better outcome, but **none of these inputs is annotated, so the counts cannot establish that the merged peaks are one compound.** I am not proposing a new default on this evidence.

For reference, MZmine's local minimum resolver exposes the equivalent quantity as `SEARCH_RT_RANGE` and its help says to *"Start optimizing with a value close to the FWHM of a peak"* — i.e. roughly a full FWHM, not a floored half of it, and user-settable rather than derived.

## What I would suggest

1. The half-window should be derived so that `chrom_fwhm` actually influences it across its useful range, rather than collapsing to 1 or 0 through `win_size / 2`.
2. The degenerate half-window of 0 should not silently accept every point as a maximum.
3. Whether the default should also widen is a separate question that needs annotated data to answer, and should not ride on (1) and (2).

Two things this investigation ruled out, so they do not need revisiting: the `min_int` clamp to `>= 1.0` is never active (0 of 250 accepted splits on `FeatureFinderMetabo_6_input.mzML`), and the `>= 2.0` valley-depth requirement is already stricter than the comparable tools — MZmine's `MIN_RATIO` defaults to 1.7 and XCMS's `MergeNeighboringPeaksParam` keeps peaks separate down to an effective ratio of 1.33.

---
_Generated by [Claude Code](https://claude.ai/code)_
